src/ride/track_design.c

```diff
@@ -119,187 +115,264 @@ rct_track_td6 *track_design_open(const utf8 *path)
 	return NULL;
 }
 
-static rct_track_td6 *track_design_open_from_buffer(uint8 *src, size_t srcLength)
+static rct_track_td6 * track_design_open_from_td4(uint8 *src, size_t srcLength)
 {
-	rct_track_td6 *td6 = calloc(1, sizeof(rct_track_td6));
-	uint8 *readPtr = src;
-
-	// Read start of track_design
-	copy(td6, &readPtr, 32);
+	rct_track_td4 * td4 = calloc(1, sizeof(rct_track_td4));
+	if (td4 == NULL) {
+		log_error("Unable to allocate memory for TD4 data.");
+		SafeFree(td4);
+		return NULL;
+	}
 
-	uint8 version = td6->version_and_colour_scheme >> 2;
-	if (version > 2) {
+	uint8 version = (src[7] >> 2) & 2;
+	if (version == 0) {
+		memcpy(td4, src, 0x38);
+		td4->elementsSize = srcLength - 0x38;
+		td4->elements = malloc(td4->elementsSize);
+		if (td4->elements == NULL) {
+			log_error("Unable to allocate memory for TD4 element data.");
+			SafeFree(td4);
+			return NULL;
+		}
+		memcpy(td4->elements, src + 0x38, td4->elementsSize);
+	} else if (version == 1) {
+		memcpy(td4, src, 0xC4);
+		td4->elementsSize = srcLength - 0xC4;
+		td4->elements = malloc(td4->elementsSize);
+		if (td4->elements == NULL) {
+			log_error("Unable to allocate memory for TD4 element data.");
+			SafeFree(td4);
+			return NULL;
+		}
+		memcpy(td4->elements, src + 0xC4, td4->elementsSize);
+	} else {
 		log_error("Unsupported track design.");
-		free(td6);
+		SafeFree(td4);
 		return NULL;
 	}
 
-	// In TD6 there are 32 sets of two byte vehicle colour specifiers
-	// In TD4 there are 12 sets so the remaining 20 need to be read
-	if (version == 2) {
-		copy(&td6->vehicle_colours[12], &readPtr, 40);
+	rct_track_td6 * td6 = calloc(1, sizeof(rct_track_td6));
+	if (td6 == NULL) {
+		log_error("Unable to allocate memory for TD6 data.");
+		SafeFree(td4);
+		return NULL;
+	}
+	
+	// Convert RCT1 ride type to RCT2 ride type
+	uint8 rct1RideType = td4->type;
+	if (td4->type == RCT1_RIDE_TYPE_WOODEN_ROLLER_COASTER) {
+		td6->type = RIDE_TYPE_WOODEN_ROLLER_COASTER;
+		td6->ride_mode = td4->mode;
+	} else if (td4->type == RCT1_RIDE_TYPE_STEEL_CORKSCREW_ROLLER_COASTER) {
```

Comment:
The ride type does not get set if the ride is a steel corkscrew because of this clause. I would just get rid of the clause, it's there to prevent using hypercoaster trains in a way that was possible in RCT1 but not in RCT2. Since we're aiming for RCT1 compatibility, we should allow it.